sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,236 @@
+                    HasCompleted = true;
+
+                    throw OperationFailedException;
+                }
+            }
+        }
+
+        private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            using DiagnosticScope scope = ClientDiagnostics.CreateScope($"{OperationTypeName}.UpdateStatus");
+
+            foreach (KeyValuePair<string, string> attribute in ScopeAttributes)
+            {
+                scope.AddAttribute(attribute.Key, attribute.Value);
+            }
+
+            scope.Start();
+
+            try
+            {
+                return await UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
+            }
+            catch (RequestFailedException e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+            catch (Exception e)
+            {
+                var requestFailedException = new RequestFailedException("Unexpected failure.", e);
+                scope.Failed(requestFailedException);
+                throw requestFailedException;
+            }
+        }
+    }
+
+#pragma warning disable SA1402 // File may only contain a single type
+    internal class OperationInternal<TResult, TResponseType> : OperationInternal<TResult>
+#pragma warning restore SA1402 // File may only contain a single type
+    {
+        private readonly IOperation<TResult, TResponseType> _operation;
+
+        public OperationInternal(ClientDiagnostics clientDiagnostics, IOperation<TResult, TResponseType> operation) : base(clientDiagnostics)
+        {
+            _operation = operation;
+            OperationTypeName = operation.GetType().Name;
+        }
+
+        protected override async ValueTask<Response> UpdateStateAsync(bool async, CancellationToken cancellationToken)
+        {
+            Response<TResponseType> response = async
+                ? await _operation.GetResponseAsync(cancellationToken).ConfigureAwait(false)
+                : _operation.GetResponse(cancellationToken);
+
+            var state = _operation.UpdateState(response);
+            var rawResponse = response.GetRawResponse();
+
+            SetState(state, rawResponse);
+
+            return rawResponse;
+        }
+    }
+
+    internal interface IOperation<TResult>
+    {
+        Task<Response> GetResponseAsync(CancellationToken cancellationToken);
+
+        Response GetResponse(CancellationToken cancellationToken);
+
+        OperationState<TResult> UpdateState(Response response);
+    }
+
+    internal interface IOperation<TResult, TResponseType>
+    {
+        Task<Response<TResponseType>> GetResponseAsync(CancellationToken cancellationToken);
+
+        Response<TResponseType> GetResponse(CancellationToken cancellationToken);
+
+        OperationState<TResult> UpdateState(Response<TResponseType> response);
+    }
+
+    internal struct OperationState<TResult>
```

Comment:
Let's make this one a readonly struct. Mutable structs have some semantics that can be surprising for implementers.

Reply:
I would add a few static method to make the construction easier. 

```C#
if (status == OperationStatus.Succeeded)
            {
                return OperationState.Success(ConvertValue(response.Value.AnalyzeResult.PageResults, response.Value.AnalyzeResult.ReadResults));
            }
```

Reply:
Updated. Added static methods `Success` and `Failure`. In case of a pending operation, `default` is expected to be returned.